keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -2641,42 +2705,36 @@ when the feature is disabled.
 
 ### Rollout, Upgrade and Rollback Planning
 
-<!--
-This section must be completed when targeting beta to a release.
--->
-
 ###### How can a rollout or rollback fail? Can it impact already running workloads?
 
-<!--
-Try to be as paranoid as possible - e.g., what if some components will restart
-mid-rollout?
+Workloads not using ResourceClaims should not be impacted because the new code
+will not do anything besides checking the Pod for ResourceClaims.
 
-Be sure to consider highly-available clusters, where, for example,
-feature flags will be enabled on some API servers and not others during the
-rollout. Similarly, consider large clusters and how enablement/disablement
-will rollout across nodes.
--->
+When kube-controller-manager fails to create ResourceClaims from
+ResourceClaimTemplates, those Pods will not get scheduled. Bugs in
+kube-scheduler might lead to not scheduling Pods that could run or worse,
+schedule Pods that should not run. Those then will get stuck on a node where
+kubelet will refuse to start them. None of these scenarios affect already
+running workloads.
+
+Failures in kubelet might affect running workloads, but only if containers for
+those workloads need to be restarted.
 
 ###### What specific metrics should inform a rollback?
 
-<!--
-What signals should users be paying attention to when the feature is young
-that might indicate a serious problem?
--->
+One indicator are unexpected restarts of the cluster control plane
```

Comment:
I think you should answer the question with specific metrics.

Reply:
Not sure. The comment says "what signals should users be paying attention" - those signals doe not have to be metrics.

Reply:
The question is "what specific metrics should inform a rollback"?

Reply:
I'll leave this for PRR reviewer/approver. Typically people would mention specific metrics that you could view to be informed of a rollback. But I also see some mentions of logs or events that could be emitted.

Reply:
Specific metrics are definitely better than general statements, whenever possible.

For restarts, which components specifically? api server, controller manager, scheduler?

What about scheduling failures? We are executing new code in scheduler, it shouldn't come into play unless they are using claims, but, you know, bugs happen. So, are there specific scheduling metrics that should be observed? Anything from the kubelet on errors related to calls to DRA drivers?

Reply:
> For restarts, which components specifically? api server, controller manager, scheduler?

All of those. Added, just for clarity.

> What about scheduling failures?

Failures in which sense? kube-scheduler cannot distinguish between "pod cannot be scheduled" and "scheduler plugin is broken".

> So, are there specific scheduling metrics that should be observed?

I've added something about the `scheduler_pending_pods` metric - perhaps that is what you are looking for.

> Anything from the kubelet on errors related to calls to DRA drivers?

Not yet. FWIW, nothing like that was added for CSI or device plugins either. @bart0sh proposed adding something in https://github.com/kubernetes/kubernetes/pull/127146 but what specifically should be tracked and whether it should be tracked by DRA driver needs further discussion.

`kubelet_started_containers_errors_total` might be useful to detect problems in kubelet. Added.

Reply:
> nothing like that was added for CSI or device plugins either

I was mistaken there. There is `csi_operations_seconds`.

Perhaps we should be using `dra_operations_seconds`?

Reply:
> Perhaps we should be using dra_operations_seconds?

Can't the calls return errors? seconds are nice but error count seems more important to me. So maybe the pair 
of errors and latency, or if we want to be really complete: total count of calls, latency, error count. I would ask SIG instrumentation if they prefer a metric per call or a histogram with method names.

Reply:
@pohly This makes sense to me. I'll add this metric to [my PR](https://github.com/kubernetes/kubernetes/pull/127146).

Reply:
@johnbelamaric: with the ` grpc_status_code`, you automatically get error counts because each histogram with a certain label also represents the total number of calls which returned that error code. If someone wants "calls with *any* error", then they can aggregate.

Reply:
BTW, I was also considering whether "number of claims" should be another label. The duration of a gRPC might depend on how much work it has to do, which can skew the histogram if we combine calls with one claim and those which, say, have 10. But I think that's too detailed.

Reply:
I agree, number of claims is too detailed.